tests/functional/test_s3.py

```diff
@@ -1454,30 +1454,61 @@ def setUp(self):
     def get_sent_headers(self):
         return self.http_stubber.requests[0].headers
 
-    def test_content_md5_set(self):
+    def test_trailing_checksum_set(self):
         with self.http_stubber:
             self.client.put_object(Bucket="foo", Key="bar", Body="baz")
-        self.assertIn("content-md5", self.get_sent_headers())
+        sent_headers = self.get_sent_headers()
+        self.assertEqual(sent_headers["Content-Encoding"], b"aws-chunked")
+        self.assertEqual(sent_headers["Transfer-Encoding"], b"chunked")
+        self.assertEqual(
+            sent_headers["X-Amz-Trailer"], b"x-amz-checksum-crc32"
+        )
+        self.assertEqual(sent_headers["X-Amz-Decoded-Content-Length"], b"3")
+        self.assertEqual(
+            sent_headers["x-amz-content-sha256"],
+            b"STREAMING-UNSIGNED-PAYLOAD-TRAILER",
+        )
 
-    def test_content_md5_set_empty_body(self):
+    def test_trailing_checksum_set_empty_body(self):
         with self.http_stubber:
             self.client.put_object(Bucket="foo", Key="bar", Body="")
-        self.assertIn("content-md5", self.get_sent_headers())
+        sent_headers = self.get_sent_headers()
+        self.assertEqual(sent_headers["Content-Encoding"], b"aws-chunked")
+        self.assertEqual(sent_headers["Transfer-Encoding"], b"chunked")
+        self.assertEqual(
+            sent_headers["X-Amz-Trailer"], b"x-amz-checksum-crc32"
+        )
+        self.assertEqual(sent_headers["X-Amz-Decoded-Content-Length"], b"0")
+        self.assertEqual(
+            sent_headers["x-amz-content-sha256"],
+            b"STREAMING-UNSIGNED-PAYLOAD-TRAILER",
+        )
 
-    def test_content_md5_set_empty_file(self):
+    def test_trailing_checksum_set_empty_file(self):
         with self.http_stubber:
             with temporary_file("rb") as f:
                 assert f.read() == b""
                 self.client.put_object(Bucket="foo", Key="bar", Body=f)
-        self.assertIn("content-md5", self.get_sent_headers())
+        sent_headers = self.get_sent_headers()
+        self.assertEqual(sent_headers["Content-Encoding"], b"aws-chunked")
+        self.assertEqual(sent_headers["Transfer-Encoding"], b"chunked")
+        self.assertEqual(
+            sent_headers["X-Amz-Trailer"], b"x-amz-checksum-crc32"
+        )
+        self.assertEqual(sent_headers["X-Amz-Decoded-Content-Length"], b"0")
+        self.assertEqual(
+            sent_headers["x-amz-content-sha256"],
+            b"STREAMING-UNSIGNED-PAYLOAD-TRAILER",
+        )
 
-    def test_content_sha256_set_if_config_value_is_true(self):
-        # By default, put_object() does not include an x-amz-content-sha256
-        # header because it also includes a `Content-MD5` header. The
-        # `payload_signing_enabled` config overrides this logic and forces the
-        # header.
+    def test_content_sha256_not_set_if_config_value_is_true(self):
+        # By default, put_object() provides a trailing checksum and includes the
```

Comment:
Is this right? I wouldn't expect us to be using trailing checksums by default on `put_object`.

Reply:
Because CRC32 is the new default, `put_object` will use trailing checksums by default. This is because `put_object` has streaming input.

You can see this behavior in the current version of the SDK by running:
```
import boto3
boto3.set_stream_logger('')

s3 = boto3.client("s3")

response = s3.put_object(
    Body=b"Example data.",
    Bucket="<bucket_name>",
    Key="example_key.txt",
    ChecksumAlgorithm="CRC32",
)
```